File: alt/ps2logo.c
/*
 * Copyright (C) 2006-2013 Mathias Lafeldt <mathias.lafeldt@gmail.com>
 *
 * This file is part of ps2logo, the PS2 logo decryptor.
 *
 * ps2logo is licensed under the terms of the MIT License. See LICENSE file.
 */

#include <stdio.h>
#include <stdlib.h>

/* Size of PS2 logo in bytes: 12 sectors x 2048 bytes */
#define LOGO_SIZE	(12 * 2048)


#if 0
static void encrypt_logo(unsigned char *logo, unsigned char key)
{
	int i;

	for (i = 0; i < LOGO_SIZE; i++)
		logo[i] = ((logo[i] << 5) | (logo[i] >> 3)) ^ key;
}
#endif

static void decrypt_logo(unsigned char *logo)
{
	unsigned char key = logo[0];
	int i;

	for (i = 0; i < LOGO_SIZE; i++) {
		logo[i] ^= key;
		logo[i] = (logo[i] << 3) | (logo[i] >> 5);
	}
}

int main(int argc, char *argv[])
{
	FILE *fp = NULL;
	unsigned char logo[LOGO_SIZE];

	if (argc < 3) {
		fprintf(stderr, "usage: %s <input file> <output file>\n", argv[0]);
		return EXIT_FAILURE;
	}

	fp = fopen(argv[1], "rb");
	if (fp == NULL) {
		fprintf(stderr, "Error: could not open input file '%s'\n", argv[1]);
		return EXIT_FAILURE;
	}

	if (fread(logo, LOGO_SIZE, 1, fp) != 1) {
		fprintf(stderr, "Error: could not read logo from input file\n");
		fclose(fp);
		return EXIT_FAILURE;
	}

	fclose(fp);
	fp = fopen(argv[2], "wb");
	if (fp == NULL) {
		fprintf(stderr, "Error: could not open output file '%s'\n", argv[2]);
		return EXIT_FAILURE;
	}

	decrypt_logo(logo);
	fwrite(logo, LOGO_SIZE, 1, fp);
	fclose(fp);

	return EXIT_SUCCESS;
}

File: alt/ps2logo.rs
use std::env;
use std::fs::File;
use std::io;
use std::io::prelude::*;
use std::path::Path;
use std::process;

fn main() {
    let args: Vec<String> = env::args().skip(1).collect();
    if args.len() != 2 {
        eprintln!("usage: ps2logo <input file> <output file>");
        process::exit(1);
    }

    let mut logo = vec![0xca; LOGO_SIZE].into_boxed_slice();

    if let Err(err) = read_logo(Path::new(&args[0]), &mut logo) {
        eprintln!("error: failed to read logo: {}", err);
        process::exit(1);
    }

    decrypt_logo(&mut logo);

    if let Err(err) = write_logo(Path::new(&args[1]), &logo) {
        eprintln!("error: failed to write logo: {}", err);
        process::exit(1);
    }
}

fn read_logo(path: &Path, buf: &mut [u8]) -> Result<(), io::Error> {
    File::open(path)?.read_exact(buf)
}

fn write_logo(path: &Path, buf: &[u8]) -> Result<(), io::Error> {
    File::create(path)?.write_all(buf)
}

fn decrypt_logo(logo: &mut [u8]) {
    let key = logo[0];
    for x in logo {
        *x ^= key;
        *x = (*x << 3) | (*x >> 5);
    }
}

// Size of PS2 logo in bytes: 12 sectors x 2048 bytes
const LOGO_SIZE: usize = 12 * 2048;
